File: ios/App/VerbaWidgets/Textures.swift
import CoreGraphics

/// The procedural half of src/finishes/catalog.ts. Nine of the sixteen finishes are not gradients
/// at all — the web grows them from SVG feTurbulence mapped through a palette ramp, which is what
/// makes steel look brushed and rust look oxidised. Core Graphics has no turbulence primitive, so
/// the same shape is generated here: fractal value noise through the identical piecewise-linear
/// ramp. The hash is the Android widget's, bit for bit, so both platforms grow the same material.
///
/// Frequencies are the catalog's own baseFrequency values — cycles per user unit on its 900-unit
/// tile. Every material's structure is low-frequency (3-8 cycles) and survives at widget size; the
/// grain and brushing layers are far finer than a widget pixel and land as soft noise instead of
/// crisp speckle. Octaves below a pixel are skipped, since they would only cost time.
enum Textures {
    private static let tile: Float = 900
    /// Bilinear value noise clusters tighter around its mean than Perlin does; widen to match.
    private static let contrast: Float = 1.3
    /// Turning each octave off-axis hides the lattice that bilinear noise would otherwise show.
    private static let octaveTurn: Float = 31
    /// Below a pixel per cycle the noise is sampled faster than it can be interpolated, so it
    /// arrives as uniform randomness — already wider than Perlin, and it needs narrowing rather
    /// than the widening interpolated noise wants. This is what keeps grain from turning to grit.
    private static let grainContrast: Float = 0.7

    /// Every colour here came from CSS, so it is sRGB and must be named as such. Core Graphics
    /// otherwise reaches for Generic RGB, whose gamma is 1.8 — enough to read a midtone about 19
    /// units light once it is re-encoded for the screen.
    static let colorSpace = CGColorSpace(name: CGColorSpace.sRGB) ?? CGColorSpaceCreateDeviceRGB()

    /// One feTurbulence call: frequency per axis, octaves, seed, and the tile's rotation.
    struct Noise {
        let freqX: Float
        let freqY: Float
        let octaves: Int
        let seed: Int
        let rotate: Float
        let turbulence: Bool

        init(freqX: Float, freqY: Float, octaves: Int, seed: Int,
             rotate: Float = 0, turbulence: Bool = false) {
            self.freqX = freqX
            self.freqY = freqY
            self.octaves = octaves
            self.seed = seed
            self.rotate = rotate
            self.turbulence = turbulence
        }
    }

    /// feColorMatrix + feComponentTransfer: one noise channel read through a colour ramp.
    static func palette(_ side: Int, _ spec: Noise, _ ramp: [UInt32]) -> CGImage? {
        let values = field(side, side, spec)
        return image(side) { pixels in
            for i in 0..<values.count {
                pixels[i] = 0xFF00_0000 | colour(ramp, values[i])
            }
        }
    }

    /// A flat colour whose alpha comes off the noise — the catalog's cloud, speck and grain layers.
    static func wash(_ side: Int, _ spec: Noise, _ rgb: UInt32, _ alphas: [Float]) -> CGImage? {
        let values = field(side, side, spec)
        return image(side) { pixels in
            for i in 0..<values.count {
                let alpha = UInt32((table(alphas, values[i]) * 255).rounded())
                // Core Graphics wants its alpha premultiplied, unlike Android's Bitmap
                let red = ((rgb >> 16) & 0xFF) * alpha / 255
                let green = ((rgb >> 8) & 0xFF) * alpha / 255
                let blue = (rgb & 0xFF) * alpha / 255
                pixels[i] = (alpha << 24) | (red << 16) | (green << 8) | blue
            }
        }
    }

    /// 0xAARRGGBB per pixel, which is what byteOrder32Little + premultipliedFirst spells on a
    /// little-endian machine — the same word layout the Android port writes.
    private static func image(_ side: Int, _ fill: (UnsafeMutablePointer<UInt32>) -> Void) -> CGImage? {
        let info = CGImageAlphaInfo.premultipliedFirst.rawValue | CGBitmapInfo.byteOrder32Little.rawValue
        guard let context = CGContext(
            data: nil, width: side, height: side, bitsPerComponent: 8, bytesPerRow: side * 4,
            space: colorSpace, bitmapInfo: info
        ), let data = context.data else { return nil }

        fill(data.bindMemory(to: UInt32.self, capacity: side * side))
        return context.makeImage()
    }

    private static func field(_ width: Int, _ height: Int, _ spec: Noise) -> [Float] {
        let cyclesX = spec.freqX * tile
        let cyclesY = spec.freqY * tile
        let octaves = octaveCount(spec.octaves, max(cyclesX, cyclesY), max(width, height))
        let radians = spec.rotate * .pi / 180
        let cosine = cos(radians)
        let sine = sin(radians)
        let stretch = max(cyclesX, cyclesY) >= Float(max(width, height)) / 2 ? grainContrast : contrast
        // Streaked finishes must keep their direction, so only isotropic noise gets turned
        let turn = spec.freqX == spec.freqY
        var turnCos = [Float](repeating: 1, count: octaves)
        var turnSin = [Float](repeating: 0, count: octaves)
        for octave in 0..<octaves {
            let angle = (turn ? octaveTurn * Float(octave) : 0) * .pi / 180
            turnCos[octave] = cos(angle)
            turnSin[octave] = sin(angle)
        }

        var out = [Float](repeating: 0, count: width * height)
        var mean: Float = 0
        for y in 0..<height {
            let v = Float(y) / Float(height)
            for x in 0..<width {
                let u = Float(x) / Float(width)
                let sx = spec.rotate == 0 ? u : u * cosine - v * sine
                let sy = spec.rotate == 0 ? v : u * sine + v * cosine
                var sum: Float = 0
                var total: Float = 0
                var amplitude: Float = 1
                var step: Float = 1
                for octave in 0..<octaves {
                    let ox = (sx * turnCos[octave] - sy * turnSin[octave]) * cyclesX * step
                    let oy = (sx * turnSin[octave] + sy * turnCos[octave]) * cyclesY * step
                    let n = noise(ox, oy, spec.seed + octave * 101)
                    sum += amplitude * (spec.turbulence ? abs(2 * n - 1) : n)
                    total += amplitude
                    amplitude *= 0.5
                    step *= 2
                }
                let value = sum / total
                out[y * width + x] = value
                mean += value
            }
        }

        // Stretch about the field's own mean, not about 0.5: folding for turbulence pulls the mean
        // well below centre, and stretching about centre would then drag the whole material dark
        mean /= Float(out.count)
        for i in 0..<out.count {
            out[i] = clamp(mean + (out[i] - mean) * stretch)
        }
        return out
    }

    /// Octaves whose cycles fall below two pixels add nothing a widget can show.
    private static func octaveCount(_ wanted: Int, _ cycles: Float, _ side: Int) -> Int {
        var limit = 1
        while limit < wanted && cycles * Float(1 << limit) < Float(side) / 2 { limit += 1 }
        return limit
    }

    private static func noise(_ x: Float, _ y: Float, _ seed: Int) -> Float {
        let x0 = Int32(x.rounded(.down))
        let y0 = Int32(y.rounded(.down))
        let fx = smooth(x - x.rounded(.down))
        let fy = smooth(y - y.rounded(.down))
        let seed32 = Int32(truncatingIfNeeded: seed)
        let top = lerp(hash(x0, y0, seed32), hash(x0 + 1, y0, seed32), fx)
        let bottom = lerp(hash(x0, y0 + 1, seed32), hash(x0 + 1, y0 + 1, seed32), fx)
        return lerp(top, bottom, fy)
    }

    /// Unsigned throughout: Java's int wraps on overflow and its >>> is a logical shift, so this
    /// is the same arithmetic the Android widget does, and it lands on the same bits.
    private static func hash(_ x: Int32, _ y: Int32, _ seed: Int32) -> Float {
        var h = UInt32(bitPattern: x) &* 374_761_393
        h = h &+ UInt32(bitPattern: y) &* 668_265_263
        h = h &+ UInt32(bitPattern: seed) &* 1_274_126_177
        h = (h ^ (h >> 13)) &* 1_274_126_177
        return Float((h ^ (h >> 16)) & 0xFF_FFFF) / Float(0xFF_FFFF)
    }

    private static func smooth(_ t: Float) -> Float {
        t * t * (3 - 2 * t)
    }

    private static func lerp(_ a: Float, _ b: Float, _ t: Float) -> Float {
        a + (b - a) * t
    }

    private static func clamp(_ v: Float) -> Float {
        v < 0 ? 0 : (v > 1 ? 1 : v)
    }

    /// feComponentTransfer type='table': the value indexes a piecewise-linear list of stops.
    private static func table(_ values: [Float], _ v: Float) -> Float {
        let position = clamp(v) * Float(values.count - 1)
        let index = min(Int(position), values.count - 2)
        return lerp(values[index], values[index + 1], position - Float(index))
    }

    private static func colour(_ colours: [UInt32], _ v: Float) -> UInt32 {
        let position = clamp(v) * Float(colours.count - 1)
        let index = min(Int(position), colours.count - 2)
        let t = position - Float(index)
        let from = colours[index]
        let to = colours[index + 1]
        let red = UInt32(lerp(Float((from >> 16) & 0xFF), Float((to >> 16) & 0xFF), t).rounded())
        let green = UInt32(lerp(Float((from >> 8) & 0xFF), Float((to >> 8) & 0xFF), t).rounded())
        let blue = UInt32(lerp(Float(from & 0xFF), Float(to & 0xFF), t).rounded())
        return (red << 16) | (green << 8) | blue
    }
}

File: android/app/src/main/java/com/verba/clock/widget/Textures.java
package com.verba.clock.widget;

import android.graphics.Bitmap;

/**
 * The procedural half of src/finishes/catalog.ts. Nine of the sixteen finishes are not gradients
 * at all — the web grows them from SVG feTurbulence mapped through a palette ramp, which is what
 * makes steel look brushed and rust look oxidised. Skia has no turbulence primitive, so the same
 * shape is generated here: fractal value noise through the identical piecewise-linear ramp.
 *
 * Frequencies are the catalog's own baseFrequency values — cycles per user unit on its 900-unit
 * tile. Every material's structure is low-frequency (3-8 cycles) and survives at widget size;
 * the grain and brushing layers are far finer than a widget pixel and land as soft noise instead
 * of crisp speckle. Octaves below a pixel are skipped, since they would only cost time.
 */
final class Textures {
    private static final float TILE = 900f;
    /** Bilinear value noise clusters tighter around its mean than Perlin does; widen to match. */
    private static final float CONTRAST = 1.3f;
    /** Turning each octave off-axis hides the lattice that bilinear noise would otherwise show. */
    private static final float OCTAVE_TURN = 31f;
    /**
     * Below a pixel per cycle the noise is sampled faster than it can be interpolated, so it
     * arrives as uniform randomness — already wider than Perlin, and it needs narrowing rather
     * than the widening interpolated noise wants. This is what keeps grain from turning to grit.
     */
    private static final float GRAIN_CONTRAST = 0.7f;

    /** One feTurbulence call: frequency per axis, octaves, seed, and the tile's rotation. */
    static final class Noise {
        private final float freqX;
        private final float freqY;
        private final int octaves;
        private final int seed;
        private final float rotate;
        private final boolean turbulence;

        Noise(float freqX, float freqY, int octaves, int seed, float rotate, boolean turbulence) {
            this.freqX = freqX;
            this.freqY = freqY;
            this.octaves = octaves;
            this.seed = seed;
            this.rotate = rotate;
            this.turbulence = turbulence;
        }
    }

    /** feColorMatrix + feComponentTransfer: one noise channel read through a colour ramp. */
    static Bitmap palette(int width, int height, Noise spec, int[] ramp) {
        float[] field = field(width, height, spec);
        int[] pixels = new int[field.length];
        for (int i = 0; i < field.length; i++) {
            pixels[i] = 0xFF000000 | ramp(ramp, field[i]);
        }
        return Bitmap.createBitmap(pixels, width, height, Bitmap.Config.ARGB_8888);
    }

    /** A flat colour whose alpha comes off the noise — the catalog's cloud and speck layers. */
    static Bitmap wash(int width, int height, Noise spec, int rgb, float[] alphas) {
        float[] field = field(width, height, spec);
        int[] pixels = new int[field.length];
        for (int i = 0; i < field.length; i++) {
            int alpha = Math.round(table(alphas, field[i]) * 255f);
            pixels[i] = (alpha << 24) | rgb;
        }
        return Bitmap.createBitmap(pixels, width, height, Bitmap.Config.ARGB_8888);
    }

    private Textures() {}

    private static float[] field(int width, int height, Noise spec) {
        float cyclesX = spec.freqX * TILE;
        float cyclesY = spec.freqY * TILE;
        int octaves = octaves(spec.octaves, Math.max(cyclesX, cyclesY), Math.max(width, height));
        double radians = Math.toRadians(spec.rotate);
        float cos = (float) Math.cos(radians);
        float sin = (float) Math.sin(radians);
        float contrast = Math.max(cyclesX, cyclesY) >= Math.max(width, height) / 2f
            ? GRAIN_CONTRAST
            : CONTRAST;
        // Streaked finishes must keep their direction, so only isotropic noise gets turned
        boolean turn = spec.freqX == spec.freqY;
        float[] turnCos = new float[octaves];
        float[] turnSin = new float[octaves];
        for (int octave = 0; octave < octaves; octave++) {
            double angle = Math.toRadians(turn ? OCTAVE_TURN * octave : 0);
            turnCos[octave] = (float) Math.cos(angle);
            turnSin[octave] = (float) Math.sin(angle);
        }

        float[] out = new float[width * height];
        float mean = 0;
        for (int y = 0; y < height; y++) {
            float v = (float) y / height;
            for (int x = 0; x < width; x++) {
                float u = (float) x / width;
                float sx = spec.rotate == 0 ? u : u * cos - v * sin;
                float sy = spec.rotate == 0 ? v : u * sin + v * cos;
                float sum = 0;
                float total = 0;
                float amplitude = 1;
                float step = 1;
                for (int octave = 0; octave < octaves; octave++) {
                    float ox = (sx * turnCos[octave] - sy * turnSin[octave]) * cyclesX * step;
                    float oy = (sx * turnSin[octave] + sy * turnCos[octave]) * cyclesY * step;
                    float n = noise(ox, oy, spec.seed + octave * 101);
                    sum += amplitude * (spec.turbulence ? Math.abs(2 * n - 1) : n);
                    total += amplitude;
                    amplitude *= 0.5f;
                    step *= 2;
                }
                float value = sum / total;
                out[y * width + x] = value;
                mean += value;
            }
        }

        // Stretch about the field's own mean, not about 0.5: folding for turbulence pulls the mean
        // well below centre, and stretching about centre would then drag the whole material dark
        mean /= out.length;
        for (int i = 0; i < out.length; i++) {
            out[i] = clamp(mean + (out[i] - mean) * contrast);
        }
        return out;
    }

    /** Octaves whose cycles fall below two pixels add nothing a widget can show. */
    private static int octaves(int wanted, float cycles, int side) {
        int limit = 1;
        while (limit < wanted && cycles * (1 << limit) < side / 2f) limit++;
        return limit;
    }

    private static float noise(float x, float y, int seed) {
        int x0 = (int) Math.floor(x);
        int y0 = (int) Math.floor(y);
        float fx = smooth(x - x0);
        float fy = smooth(y - y0);
        float top = lerp(hash(x0, y0, seed), hash(x0 + 1, y0, seed), fx);
        float bottom = lerp(hash(x0, y0 + 1, seed), hash(x0 + 1, y0 + 1, seed), fx);
        return lerp(top, bottom, fy);
    }

    private static float hash(int x, int y, int seed) {
        int h = x * 374_761_393 + y * 668_265_263 + seed * 1_274_126_177;
        h = (h ^ (h >>> 13)) * 1_274_126_177;
        return ((h ^ (h >>> 16)) & 0xFFFFFF) / (float) 0xFFFFFF;
    }

    private static float smooth(float t) {
        return t * t * (3 - 2 * t);
    }

    private static float lerp(float a, float b, float t) {
        return a + (b - a) * t;
    }

    private static float clamp(float v) {
        return v < 0 ? 0 : v > 1 ? 1 : v;
    }

    /** feComponentTransfer type='table': the value indexes a piecewise-linear list of stops. */
    private static float table(float[] values, float v) {
        float position = clamp(v) * (values.length - 1);
        int index = Math.min((int) position, values.length - 2);
        return lerp(values[index], values[index + 1], position - index);
    }

    private static int ramp(int[] colors, float v) {
        float position = clamp(v) * (colors.length - 1);
        int index = Math.min((int) position, colors.length - 2);
        float t = position - index;
        int from = colors[index];
        int to = colors[index + 1];
        int r = Math.round(lerp((from >> 16) & 0xFF, (to >> 16) & 0xFF, t));
        int g = Math.round(lerp((from >> 8) & 0xFF, (to >> 8) & 0xFF, t));
        int b = Math.round(lerp(from & 0xFF, to & 0xFF, t));
        return (r << 16) | (g << 8) | b;
    }
}
